test/hotspot/gtest/nmt/test_nmt_treap.cpp

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.hpp"
+#include "nmt/nmtTreap.hpp"
+#include "runtime/os.hpp"
+#include "unittest.hpp"
+
+#ifdef ASSERT
+
+TEST_VM(NmtTreap, VerifyItThroughStressTest) {
+  struct Cmp {
+    static int cmp(int a, int b) {
+      return a - b;
+    }
+  };
+  TreapCHeap<int, int,Cmp> treap;
+  // Really hammer a Treap
+  int ten_thousand = 10000;
+  for (int i = 0; i < ten_thousand; i++) {
+    int r = os::random();
+    if (r >= 0) {
+      treap.upsert(i, i);
+    } else {
+      treap.remove(i);
+    }
+  }
+  for (int i = 0; i < ten_thousand; i++) {
+    int r = os::random();
+    if (r >= 0) {
+      treap.upsert(i, i);
+    } else {
+      treap.remove(i);
+    }
+  }
```

Comment:
Please test the find functions (leq, geq etc) and iteration.

For the latter, please test that we see every value, guaranteed, and that we see every value only exactly once.

Also test:
- inserting duplicates should result in only one value
- empty treap should work as expected
- treap with custom allocator, e.g. a simple array based one. Make sure we don't leak after nodes are removed, or after remove_all
- treap with different custom comparator, please counter check the iteration order
- maybe some more types? at least the common numericals

Reply:
And I would not rely on the internal verification here, but call verify explicitly. (see also note in treap add/remove about verification)

Reply:
Please also test the scoped find function with different sets (eg. empty set, 1 item set etc).

Reply:
I've added a fair amount of tests to this code now. I still need to add some more for iterators, at least the closest_leq. Most of the tests use `int` (so not `size_t`), I wrote the `find` tests using `float` instead.

Reply:
OK, very basic closest_leq tests added. For verify_self, I've removed all calls in the treap-code itself and let a small hammering test run the verify_self code.